### PR TITLE
Mutant file formatter

### PR DESCRIPTION
**Pull request by eliottBo**
### Added

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Comment by eliottBo** [comment]
Testing in stage:
The sample file name does not contain the lab and region code.

```shell
[eliott.bosshard@hasta:~] [S_base] $ bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b mutant-file-formatter

(...)

Log deploy... done.
cg, version 64.4.14
```
```shell
[eliott.bosshard@hasta:~] [S_base] $ cg deliver case -c ethicallark -d fastq --dry-run
Running CG deliver
Called undefined __pydantic_serializer__ on HousekeeperAPI, please wrap
Called undefined __dataclass_fields__ on HousekeeperAPI, please wrap
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L005_R2_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L005_R2_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L004_R2_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L004_R2_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L001_R1_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L001_R1_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L002_R1_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L002_R1_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L008_R1_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L008_R1_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L006_R2_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L006_R2_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L001_R2_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L001_R2_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L003_R1_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L003_R1_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L008_R2_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L008_R2_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L007_R2_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L007_R2_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L007_R1_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L007_R1_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L002_R2_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L002_R2_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L005_R1_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L005_R1_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L006_R1_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L006_R1_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L003_R2_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L003_R2_001.fastq.gz
/home/proj/stage/customers/cust124/inbox/1000949/22GTNNLT4_ACC16076A1_S1_L004_R1_001.fastq.gz already exists. Overwriting with /home/proj/stage/housekeeper-bundles/ACC16076A1/2024-11-04/22GTNNLT4_ACC16076A1_S1_L004_R1_001.fastq.gz
Concatenated forward reads to /home/proj/stage/customers/cust124/inbox/1000949/24CS502234/24CS502234_1.fastq.gz
Concatenated reverse reads to /home/proj/stage/customers/cust124/inbox/1000949/24CS502234/24CS502234_2.fastq.gz
Concatenated forward reads to /home/proj/stage/customers/cust124/inbox/1000949/0PROVSEK/0PROVSEK_1.fastq.gz
Concatenated reverse reads to /home/proj/stage/customers/cust124/inbox/1000949/0PROVSEK/0PROVSEK_2.fastq.gz
Concatenated forward reads to /home/proj/stage/customers/cust124/inbox/1000949/24CS502277/24CS502277_1.fastq.gz
Concatenated reverse reads to /home/proj/stage/customers/cust124/inbox/1000949/24CS502277/24CS502277_2.fastq.gz
(...)
Concatenated forward reads to /home/proj/stage/customers/cust124/inbox/1000949/24CS502212/24CS502212_1.fastq.gz
Concatenated reverse reads to /home/proj/stage/customers/cust124/inbox/1000949/24CS502212/24CS502212_2.fastq.gz
Concatenated forward reads to /home/proj/stage/customers/cust124/inbox/1000949/24CS502279/24CS502279_1.fastq.gz
Concatenated reverse reads to /home/proj/stage/customers/cust124/inbox/1000949/24CS502279/24CS502279_2.fastq.gz
Concatenated forward reads to /home/proj/stage/customers/cust124/inbox/1000949/24CS502229/24CS502229_1.fastq.gz
Concatenated reverse reads to /home/proj/stage/customers/cust124/inbox/1000949/24CS502229/24CS502229_2.fastq.gz
Traceback (most recent call last):
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/bin/cg", line 8, in <module>
    sys.exit(base())
             ^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/click/decorators.py", line 45, in new_func
    return f(get_current_context().obj, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/cli/deliver/base.py", line 96, in deliver_case
    delivery_service.deliver_files_for_case(
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/services/deliver_files/deliver_files_service/error_handling.py", line 18, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/services/deliver_files/deliver_files_service/deliver_files_service.py", line 75, in deliver_files_for_case
    formatted_files: FormattedFiles = self.file_formatter.format_files(moved_files)
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/services/deliver_files/file_formatter/service.py", line 52, in format_files
    formatted_files: list[FormattedFile] = self._format_sample_and_case_files(
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/services/deliver_files/file_formatter/service.py", line 63, in _format_sample_and_case_files
    formatted_files: list[FormattedFile] = self.sample_file_formatter.format_files(
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py", line 25, in format_files
    formatted_files = self._add_lims_metadata(
                      ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py", line 47, in _add_lims_metadata
    lims_meta_data = self._get_lims_naming_metadata(sample_id)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py", line 32, in _get_lims_naming_metadata
    region_code = self.lims_api.get_sample_attribute(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/cg/apps/lims/api.py", line 251, in get_sample_attribute
    return sample.udf[PROP2UDF[key]]
           ^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/genologics/descriptors.py", line 346, in __get__
    instance.get()
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/genologics/entities.py", line 289, in get
    self.root = self.lims.get(self.uri)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/genologics/lims.py", line 90, in get
    return self.parse_response(r)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/genologics/lims.py", line 209, in parse_response
    self.validate_response(response, accept_status_codes)
  File "/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/genologics/lims.py", line 202, in validate_response
    raise requests.exceptions.HTTPError(message, response=response)
requests.exceptions.HTTPError: 404: Sample not found: ACC16076A1
```
```shell
[eliott.bosshard@hasta:~] [S_base] $ ll /home/proj/stage/customers/cust124/inbox/1000949/24CS502226/24CS502226_
24CS502226_1.fastq.gz  24CS502226_2.fastq.gz  
[eliott.bosshard@hasta:~] [S_base] $ ll /home/proj/stage/customers/cust124/inbox/1000949/24CS502227/24CS502227_
24CS502227_1.fastq.gz  24CS502227_2.fastq.gz 
```

**Comment by eliottBo** [comment]
Test in stage:
Files have the correct name after running `cg deliver case -c case_internal_id -d fastq --dry-run`
Screenshot 2024-11-06 at 13 35 49

File: cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    @staticmethod
+    def _get_sample_id_by_original_path(original_path: Path, sample_files: list[SampleFile]) -> str:
+        for sample_file in sample_files:
+            if sample_file.file_path == original_path:
+                return sample_file.sample_id
+        raise ValueError(f"Could not find sample file with path {original_path}")
+
+    @staticmethod
+    def _filter_unique_path_combinations(
```
Comment (islean) [inline cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py:63]: What could be the cause for duplication here? In what scenario would the original path and the formatted path be exactly the same?

File: cg/services/deliver_files/file_formatter/utils/sample_service.py [hunk 2]
```diff
@@ -28,8 +31,7 @@ def _create_sample_folders(ticket_dir_path: Path, sample_names: set[str]):
             sample_dir_path = Path(ticket_dir_path, sample_name)
             sample_dir_path.mkdir(exist_ok=True)
 
-    def _format_sample_files(self, sample_files: list[SampleFile]) -> list[FormattedFile]:
-        formatted_files: list[FormattedFile] = self._get_formatted_files(sample_files)
+    def _format_sample_files(self, formatted_files: list[FormattedFile]) -> list[FormattedFile]:
```
Comment (islean) [inline cg/services/deliver_files/file_formatter/utils/sample_service.py:34]: This naming is a bit unclearer now - I think we are using the word "format" too much. Would `rename_sample_files` (or maybe `rename_files`) be better here?

File: cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        return f"{region_code}_{lab_code}_"
+
+    def _add_lims_metadata(
+        self, formatted_files: list[FormattedFile], sample_files: list[SampleFile]
+    ) -> list[FormattedFile]:
+        for formatted_file in formatted_files:
+            sample_id: str = self._get_sample_id_by_original_path(
+                original_path=formatted_file.original_path, sample_files=sample_files
+            )
+            lims_meta_data = self._get_lims_naming_metadata(sample_id)
```
Comment (islean) [inline cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py:47]:
```suggestion
            lims_meta_data: str = self._get_lims_naming_metadata(sample_id)
```

File: cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            formatted_file.original_path = formatted_file.formatted_path
```
Comment (islean) [inline cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py:48]: This is very sneaky to me. So after calling `_add_lims_metadata`, `[file.original_path for file in formatted_files]` will be `[file.formatted_path for file in formatted_files]` for whatever is in the input?

Comment (islean) [reply cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py:48]: Maybe this ties together with what @Vince-janv mentioned when talking about BaseModels - one of the core concepts of the base model is that it is supposed to be static and not changing, and I do think that it is quite hard to follow the steps that the SampleFile and FormattedFile objects undergo in this service.

**Review by islean — COMMENTED** [review]
Adding some comments. In general I am positive, but I do think we can make the code a bit clearer since it is taking me quite some time to puzzle together what the actual flow is here. I will try to map the flow out and you can correct me. 

1. So we have SampleFiles being passed to the FileFormattingService. The first layer of formatting from the SampleFileFormatter returns a list of `FormattedFile`s which have the original path attached, and a formatted path which prepends the sample name?
2. These formatted files are now concatenated in the `SampleFileConcatenationFormatter`. This results in the formatted_path attributes becoming exchanged so that all FormattedFiles whose original_path corresponds to a file from the same sample with the same direction have the same concatenated formatted_path.
3. We now enter the MutantFileFormatter which prepends the region and region code to the file names. In doing so, all the formatted_paths from step 2 are actually set to the original paths of all the formatted files. The formatted paths are exactly the same as the original paths but with the reqion and lab code prepended.
4. We now filter out unique FormattedFiles. I guess this is because the change of the original_path in step 3 now makes it so that for a sample having 2 forward fastq files, we will have 2 duplicated FormattedFile entries? This indicates to me that we are using the list of FormattedFiles in a not intended way. 
5. We now rename all the FormattedFiles from their original path to the formatted path. Will this work? I guess the original paths all correspond to the concatenated paths now so they have been created in step 2? Is the only goal at this point to rename the already concatenated fastq files so that they contain the lims metadata?


Also - in the `_concatenate_fastq_files` in the `SampleFileConcatenationFormatter` we seem to have hard coded that we are removing the raw fastq files which I thought should not happen in this case. Is it intended that we remove the raw fastq files?

File: cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py (hunk 4 above)
Comment (ChrOertlin) [reply cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py:48]: I will re work this flow. It is indeed unclear and that has to do with the inheritance in this case.

About the basemodels I agree that they should be static, but I think that refers to their structure (i.e. the FIELDS, themselves) and not necessarily the data within those fields. As long as they are valid for what can be place on those fields.

That being said. I do think this flow suffers from re-using the same base model with the same static structure and use it in a slightly different context. I will try to rectify that.

File: cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py (hunk 1 above)
Comment (ChrOertlin) [reply cg/services/deliver_files/file_formatter/utils/mutant_sample_service.py:70]: clarified in docstring

**Review by islean — APPROVED** [review]
Looks good! Some unclarity stemming from Mutant with the multiple formatting steps but that is fine

File: cg/apps/lims/api.py [hunk 5]
```diff
@@ -555,3 +555,9 @@ def _get_negative_controls_from_list(samples: list[Sample]) -> list[Sample]:
             ):
                 negative_controls.append(sample)
         return negative_controls
+
+    def get_sample_region_and_lab_code(self, sample_id: str) -> str:
+        """Return the reqgion code and lab code for a sample formatted as a suffix string."""
+        region_code = self.get_sample_attribute(lims_id=sample_id, key="region_code").split(" ")[0]
+        lab_code = self.get_sample_attribute(lims_id=sample_id, key="lab_code").split(" ")[0]
```
Comment (ChrOertlin) [inline cg/apps/lims/api.py:562]:
```suggestion
        region_code: str = self.get_sample_attribute(lims_id=sample_id, key="region_code").split(" ")[0]
        lab_code: str = self.get_sample_attribute(lims_id=sample_id, key="lab_code").split(" ")[0]
```

File: cg/constants/constants.py [hunk 6]
```diff
@@ -100,7 +100,7 @@ class SexOptions(StrEnum):
     UNKNOWN: str = "unknown"
 
 
-SARS_COV_REGEX = "^[0-9]{2}CS[0-9]{6}$"
+SARS_COV_REGEX = "^[0-9]{2}CS[0-9]{6}TEST$"
```
Comment (ChrOertlin) [inline cg/constants/constants.py:103]:
```suggestion
SARS_COV_REGEX = "^[0-9]{2}CS[0-9]{6}$"
```

Comment (ChrOertlin) [reply cg/constants/constants.py:103]: just for testing.... easier to change the regex now so the test case can be picked up than to setup a new testing case with LIMS integration etc... we need to address this as this seems a bit weird of a check.

File: cg/services/deliver_files/deliver_files_service/deliver_files_service.py [hunk 7]
```diff
@@ -58,14 +60,23 @@ def __init__(
...
         )
         moved_files: DeliveryFiles = self.file_mover.move_files(
             delivery_files=delivery_files, delivery_base_path=delivery_base_path
         )
-        formatted_files: FormattedFiles = self.file_formatter.format_files(moved_files)
+        formatted_files: FormattedFiles = self.file_formatter.format_files(
+            delivery_files=moved_files
+        )
+        for formatted_file in formatted_files.files:
+            assert formatted_file.formatted_path.exists()
```
Comment (ChrOertlin) [inline cg/services/deliver_files/deliver_files_service/deliver_files_service.py:79]: cleanup
```suggestion
     
```

File: cg/services/deliver_files/file_formatter/component_files/abstract.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
+from abc import abstractmethod, ABC
+from pathlib import Path
+
+from cg.services.deliver_files.file_fetcher.models import SampleFile, CaseFile
+from cg.services.deliver_files.file_formatter.destination.models import FormattedFile
+
+
+class ComponentFormatter(ABC):
```
Comment (Vince-janv) [inline cg/services/deliver_files/file_formatter/component_files/abstract.py:8]: What does component mean here?

**Review by Vince-janv — APPROVED** [review]
Nice work! Since we've touched the delivery flow, have you tested that non-sars-cov2 deliveries still behave as we expect them?

File: cg/services/deliver_files/file_formatter/component_files/concatenation_service.py [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+        args:
+            concatenation_maps: list[ConcatenationMap]: List of ConcatenationMap objects.
+            formatted_files: list[FormattedFile]: List of formatted files.
+        """
+        for formatted_file in formatted_files:
+            if self._is_lane_fastq_file(formatted_file.formatted_path):
+                formatted_file.formatted_path = concatenation_maps[formatted_file.formatted_path]
+
+    @staticmethod
+    def _all_sample_fastq_file_share_same_directory(
```
Comment (diitaz93) [inline cg/services/deliver_files/file_formatter/component_files/concatenation_service.py]: The name of this method suggests a bool return, consider renaming to validate_all_fastq_files....

File: cg/services/deliver_files/file_formatter/component_files/concatenation_service.py [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+        sample_fastq_files: dict[str, list[FastqFile]]
+    ) -> None:
+        """
+        Assert that all fastq files for a sample share the same directory.
+        This is to ensure that the files are concatenated within the expected directory path.
+        args:
```
Comment (diitaz93) [inline cg/services/deliver_files/file_formatter/component_files/concatenation_service.py]: add the raise to the docstring

File: cg/services/deliver_files/file_fetcher/analysis_service.py [hunk 11]
```diff
@@ -71,9 +76,11 @@ def _validate_delivery_has_content(delivery_files: DeliveryFiles) -> DeliveryFil
     @handle_missing_bundle_errors
     def _get_sample_files_from_case_bundle(
         self, workflow: Workflow, sample_id: str, case_id: str
-    ) -> list[SampleFile]:
+    ) -> list[SampleFile] | None:
         """Return a list of files from a case bundle with a sample id as tag."""
         sample_tags: list[set[str]] = self.tags_fetcher.fetch_tags(workflow).sample_tags
+        if not sample_tags:
+            return []
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/deliver_files/file_fetcher/analysis_service.py:98]: it would return None and not an empty list otherwise. That causes issues later on.

File: cg/services/deliver_files/file_fetcher/analysis_service.py [hunk 12]
```diff
@@ -101,13 +110,17 @@ def _get_analysis_sample_delivery_files(self, case: Case) -> list[SampleFile] |
...
+    ) -> list[CaseFile] | None:
         """
         Return a complete list of analysis case files to be delivered and ignore analysis sample
         files.
         """
         case_tags: list[set[str]] = self.tags_fetcher.fetch_tags(case.data_analysis).case_tags
-        sample_id_tags: list[str] = case.sample_ids
+        if not case_tags:
+            return []
+        sample_id_tags: list[str] = [sample_id] if sample_id else case.sample_ids
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/deliver_files/file_fetcher/analysis_service.py:146]: see above

File: cg/services/deliver_files/file_formatter/component_files/concatenation_service.py [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+            sample_fastq_files: dict[str, list[FastqFile]]: Dictionary of sample names and their fastq files.
+        """
+        for sample_name in sample_fastq_files.keys():
+            fastq_files: list[FastqFile] = sample_fastq_files[sample_name]
+            parent_dir: Path = fastq_files[0].fastq_file_path.parent
+            for fastq_file in fastq_files:
+                if fastq_file.fastq_file_path.parent != parent_dir:
+                    raise ValueError(
```
Comment (diitaz93) [inline cg/services/deliver_files/file_formatter/component_files/concatenation_service.py]: Make sure this error is caught somewhere and we know what to expect when this fails. I could not see any error handling for this in the fohm upload

File: cg/services/fastq_concatenation_service/fastq_concatenation_service.py [hunk 14]
```diff
@@ -1,28 +1,48 @@
...
         fastq_directory: Path,
         forward_output_path: Path,
         reverse_output_path: Path,
         remove_raw: bool = False,
     ):
-        temp_forward: Path | None = concatenate_forward_reads(fastq_directory)
-        temp_reverse: Path | None = concatenate_reverse_reads(fastq_directory)
+        """Concatenate fastq files for a given sample in a directory and write the concatenated files to the output path.
+
+        Args:
```
Comment (diitaz93) [inline cg/services/fastq_concatenation_service/fastq_concatenation_service.py:26]: I have seen this in lower case and capitalised. Maybe standardise

File: cg/services/deliver_files/deliver_files_service/deliver_files_service.py [hunk 15]
```diff
@@ -65,7 +67,11 @@ def deliver_files_for_case(
...
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/deliver_files/deliver_files_service/deliver_files_service.py]: removed

File: cg/services/deliver_files/file_formatter/component_files/mutant_service.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+        """
+        Filter out duplicates from the formatted files list.
+
+        note:
+            During fastq concatenation Sample_R1 and Sample_R2 files are concatenated
+            and moved to the same file Concat_Sample. This mean that there can be multiple entries
+            for the same concatenated file in the formatted_files list
+            coming from the SampleFileConcatenationService.
+            This function filters out the duplicates to avoid moving the same file multiple times
+            which would result in an error the second time since the files is no longer in the original path.
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/deliver_files/file_formatter/component_files/mutant_service.py:135]: discussed in meeting yesterday

File: cg/services/deliver_files/file_formatter/component_files/abstract.py (hunk 8 above)
Comment (ChrOertlin) [reply cg/services/deliver_files/file_formatter/component_files/abstract.py:8]: discussed in meeting yesterday

File: tests/services/file_delivery/file_formatter/files/test_formatter_utils.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+            "expected_concatenated_fastq_flat_formatted_files",
+            SampleFileConcatenationFormatter(
+                file_manager=FileManager(),
+                path_name_formatter=FlatStructurePathFormatter(),
+                concatenation_service=FastqConcatenationService(),
+            ),
+        ),
+    ],
+)
+def test_component_formatters(
```
Comment (diitaz93) [inline tests/services/file_delivery/file_formatter/files/test_formatter_utils.py:73]: remember to fix the name "component" here too

**Review by diitaz93 — APPROVED** [review]
Awesome 🚀 Lets rename the component class if possible

File: cg/services/deliver_files/factory.py [hunk 18]
```diff
@@ -0,0 +1,324 @@
...
+        file_tag_fetcher: FetchDeliveryFileTagsService = self._get_file_tag_fetcher(
+            delivery_type=delivery_type, delivery_destination=delivery_destination
+        )
+        return service_map[delivery_type](
+            status_db=self.store,
+            hk_api=self.hk_api,
+            tags_fetcher=file_tag_fetcher,
+        )
+
+    def _convert_workflow(self, case: Case) -> Workflow:
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/deliver_files/factory.py:178]: We are passing all workflows to this. If it happens to be microbial a conversion happens, otherwise nothing. I think the general name is good here.

File: cg/services/fastq_concatenation_service/fastq_concatenation_service.py (hunk 14 above)
Comment (ChrOertlin) [reply cg/services/fastq_concatenation_service/fastq_concatenation_service.py:26]: I tried to fix that to all lower case. Will make sure to get them all!

File: cg/services/deliver_files/file_formatter/component_files/concatenation_service.py (hunk 13 above)
Comment (ChrOertlin) [reply cg/services/deliver_files/file_formatter/component_files/concatenation_service.py]: the whole process should fail if this error happens as it signals we are trying to do something that isn't possible. Or do you want to catch this and re-raise a more specific error?

File: cg/services/deliver_files/factory.py [hunk 19]
```diff
@@ -0,0 +1,324 @@
...
+        """Converts a workflow with the introduction of the microbial-fastq delivery type an
+        unsupported combination of delivery type and workflow setup is required. This function
+        makes sure that a raw data workflow with microbial fastq delivery type is treated as a
+        microsalt workflow so that the microbial-fastq sample files can be concatenated.
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/deliver_files/factory.py]: Sebastian rewrote it a bit I accepted his suggestion.

File: cg/services/deliver_files/deliver_files_service/deliver_files_service.py [hunk 20]
```diff
@@ -108,7 +122,29 @@ def deliver_files_for_sample(
         )
         self._add_trailblazer_tracking(case=case, job_id=job_id, dry_run=dry_run)
 
+    def deliver_files_for_fohm_upload(self, case: Case, sample_id: str, delivery_base_path: Path):
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/deliver_files/deliver_files_service/deliver_files_service.py]: I changed the function name now.

File: cg/services/deliver_files/file_formatter/files/concatenation_service.py [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+        LOG.debug(f"[FORMAT SERVICE] delivery_path: {delivery_path}")
+        fastq_files: list[FastqFile] = self._get_unique_sample_fastq_paths(
+            sample_names=sample_names, delivery_path=delivery_path
+        )
+        grouped_fastq_files: dict[str, list[FastqFile]] = self._group_fastq_files_per_sample(
+            sample_names=sample_names, fastq_files=fastq_files
+        )
+        concatenation_maps: dict[Path, Path] = {}
+        for sample in grouped_fastq_files.keys():
+            # The parent is dependent on the nested or flat structure within the delivery path.
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/deliver_files/file_formatter/files/concatenation_service.py]: I removed the comment.
